File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.cpp
/*
 * @lc app=leetcode id=141 lang=cpp
 *
 * [141] Linked List Cycle
 */
 
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode* slow = head;
        ListNode* fast = head;

        if (head == nullptr)
        {
            return false;
        }

        fast = head->next;

        while (slow != fast) {
            if (fast == nullptr)
            {
                return false;
            }

            slow = slow->next;
            fast = fast->next;

            if (fast == nullptr)
            {
                return false;
            }

            fast = fast->next;
        }

        return true;
    }
};

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.go
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	forward := head
	for i := 0; i < n; i++ {
		forward = forward.Next
	}

	prev := head
	cur := head

	for forward != nil {
		prev = cur
		cur = cur.Next
		forward = forward.Next	
	}

	if cur == head {
		return head.Next
	}

	prev.Next = cur.Next

	return head
}
